### PR TITLE
Merge 5.1.1minus 16

### PR DESCRIPTION
**Pull request by liam923**
The first four commits can be skimmed/skipped. The first commit (https://github.com/janestreet/merlin-jst/pull/65/commits/84c178d6a8ccfaa866c4610152af1693b2d7dbc8) is simply copying in the compiler code. The second commit (https://github.com/janestreet/merlin-jst/pull/65/commits/f19d0c1af15bfcd93882be1c0d736314ecb31dcb) is the result of merging in compiler changes via the import script, without any manual resolving. The third commit (https://github.com/janestreet/merlin-jst/pull/65/commits/a5021533fc975cf09b5dcc90a5b66bac69bf63e7) is manual resolution of trivial conflicts. The fourth commit (https://github.com/janestreet/merlin-jst/pull/65/commits/d31d490be052e8470fcb93537127efe70c788eca) is manual resolution of trivial type errors resulting from the merge.

The remaining commits are non-trivial (meaning they required at least some thought/decision making) and should be looked over.

### CODE REVIEW

**Review by lukemaurer — COMMENTED** [review]
`Env` and `Persistent_env` LGTM

**Comment by liam923** [comment]
The 4 more recent commits correspond to the compiler pr [2611](https://github.com/ocaml-flambda/flambda-backend/pull/2611). The first three commits of them can be skipped - they import the changes from the compiler with no manual conflict resolution.

File: src/ocaml/typing/rec_check.ml [hunk 1]
```diff
@@ -1343,7 +1343,7 @@ and is_destructuring_pattern : type k . k general_pattern -> bool =
     | Tpat_or (l,r,_) ->
         is_destructuring_pattern l || is_destructuring_pattern r
 
-let is_valid_recursive_expression idlist expr =
+let _is_valid_recursive_expression idlist expr =
```
[the remark this answers is not in the document]
Comment (goldfirere) [reply src/ocaml/typing/rec_check.ml:1346]: Or maybe the upstream-inspired changes later obviate this?

**Review by goldfirere — APPROVED** [review]

File: src/ocaml/typing/rec_check.ml (hunk 1 above)
[the remark this answers is not in the document]
Comment (liam923) [reply src/ocaml/typing/rec_check.ml]: I did this with the intention of keeping the diff small to help prevent future merge conflicts. I've just realized though that `rec_check.ml` was deleted in the compiler - unfortunately, the merge tool didn't delete the file automatically for me. I also picked up on another file that was deleted in the compiler (`type_immediacy.ml`).

**Pull request merged**